### ISSUE TITLE
2D union occasionally yields spurious gap

### ISSUE THREAD

**Issue by jordanbrown0** [openscad/openscad#5478]
**Describe the bug**
A collection of translated squares, with a couple of them touching edge-to-edge, occasionally leaves a spurious line/micro-gap between two of them.

**To Reproduce**
F6 the program below.  Note that removing any square makes the problem go away, and detaching any of the three "attached" squares makes the problem go away.

**Expected behavior**
"Good" union.

**Code reproducing the issue**
```
square([50, 3]);
translate([0, -3]) // Note this one is -3
    square(3);
translate([6, -3]) // Note this one is -3
    square(3);
translate([12, -6])
    square(3);
translate([19, -6])
    square(3);
translate([25, -6])
    square(3);
translate([31,-6])
    square(3);
translate([38,-6])
    square(3);
translate([44, -3])  // Note this one is -3
    square(3);
translate([50, -6])
    square(3);
translate([57,-6])
    square(3);
translate([63,-6])
    square(3);
translate([69,-6])
    square(3);
translate([76, -6])
    square(3);
translate([82,-6])
    square(3);
translate([88,-6])
    square(3);
translate([95,-6])
    square(3);
```

**Screenshots**
![image](https://github.com/user-attachments/assets/bd1f96ef-8e4a-4595-bf88-67f541f54525)

**Environment and Version info (please complete the following information):**
 - OS: Windows 11
 - System: Intel PC 64-bit
 - OpenSCAD Version: 2021.01, 2024.10.28

**Comment by kintel** [comment]
Since we're using floating point numbers under the hood, translations will not be exact. This is expected, but the resulting 2D object, and any subsequent 3D object should still work and be able to participate in further CSG operations.

**Comment by jordanbrown0** [comment]
Floating point is exact for reasonably small integers.  (~10 million for 32-bit float, ~10^16 for 64-bit float.)

Floating point inaccuracy might explain the problem at the one problematic join.  But it doesn't explain why deleting a square that doesn't touch the main body affects that problematic join.

**Comment by kintel** [comment]
I have a plausible explanation: The 2D union is performed using the Clipper library. Before computing the union, they quantize the space covered by the bounding box of all polygons into some sub-64-bit integer space. This means that if you change the bounds of the union, precision will change too.

**Comment by jordanbrown0** [comment]
Removing a middle square doesn't change the bounding box :-(

**Comment by kintel** [comment]
Huh, gotta find another explanation then :)

**Comment by pca006132** [comment]
Wonder if this can happen with clipper2

**Comment by kintel** [comment]
Clipper2 currently uses single precision, but otherwise looks pretty much the same I think. Single precision causes other issues which is why we haven't switched yet, but I want to try again soon

**Comment by pca006132** [comment]
I don't think clipper2 uses single precision, we use double precision with clipper2, see https://www.angusj.com/clipper2/Docs/Units/Clipper/Types/PointD.htm

**Comment by kintel** [comment]
I remembered wrong: Clipper1 used 128-bit integers in their core algos while Clipper2 uses 64-bit integers. The double interface is just the front-end before they quantize to int.

**Comment by pca006132** [comment]
but is 64 bit not enough?

**Comment by kintel** [comment]
We had a problem earlier where they used 64-bit int also for intermediate results, and something did go wrong. I believe it was fixed by using 128-bit math in some places.

Anyway, essentially, I think the net effect is that Clipper2 will be more aggressive at merging vertices than Clipper1.

**Comment by pca006132** [comment]
That sounds like a bug in clipper rather than a precision issue...

**Comment by kintel** [comment]
Yeah, but I think they were unable to address it in other ways than just throwing bits at it..

**Comment by pca006132** [comment]
I see,. It would be nice if you can dig out the issue and see if clipper2 is fine with that, and file a report if it does not work.

**Comment by kintel** [comment]
Related: https://github.com/AngusJohnson/Clipper2/issues/777

**Comment by kintel** [comment]
Turns out we don't yet use the Manifold triangulator, so I think this is actually an issue with our existing triangulator.
Perhaps a good time to attempt a switch again. I believe all issues we opened has been fixed, I just haven't had time to go through all the torture tests..

**Comment by kintel** [comment]
Related to #5482

**Comment by kintel** [comment]
Btw., I just realized I cannot reproduce this; it might come down to CPU differences..

**Comment by pca006132** [comment]
Reproduced on Linux with an Intel and AMD CPU, so yeah, sounds like either due to mac or arm...

**Comment by pca006132** [comment]
Wonder what if you compile for x86 and run it with rosetta?

**Comment by kintel** [comment]
Humm, not with Rosetta either.

Anyway, I just merged #5483 which fixes the upstream issue which is about making sure extruded polygons like this become manifold in 3D space.

But we can keep this open until we upgrade to Clipper2 and see if that changes the behavior of this.

**Comment by kintel** [comment]
@jordanbrown0 Could you give it a spin with the latest master, where Clipper2 support was merged?

**Comment by jordanbrown0** [comment]
Sorry, dropped on the floor.

Not duplicated with 2026.04.26.

**Comment by jordanbrown0** [comment]
Quite possibly a duplicate of #2413.

**Comment by jordanbrown0** [comment]
Let's call it fixed.
